Source/Core/Core/HW/DVDThread.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <cinttypes>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "Common/ChunkFile.h"
+#include "Common/CommonTypes.h"
+#include "Common/Event.h"
+#include "Common/Flag.h"
+#include "Common/MsgHandler.h"
+#include "Common/Thread.h"
+#include "Common/Timer.h"
+#include "Common/Logging/Log.h"
+
+#include "Core/Core.h"
+#include "Core/CoreTiming.h"
+#include "Core/HW/DVDInterface.h"
+#include "Core/HW/DVDThread.h"
+#include "Core/HW/Memmap.h"
+#include "Core/HW/SystemTimers.h"
+
+#include "DiscIO/Volume.h"
+
+namespace DVDThread
+{
+
+static void DVDThread();
+static bool ReadFromDisc();
+
+static void FinishRead(u64 userdata, int cyclesLate);
+static int s_finish_read;
+
+static std::thread s_DVD_thread;
+static Common::Event s_DVD_thread_start_working;
+static Common::Event s_DVD_thread_not_working;
+static Common::Flag s_DVD_thread_exiting(false);
+
+static std::vector<u8> s_DVD_buffer;
+static u64 s_time_read_started;
+static bool s_DVD_success;
+
+static u64 s_DVD_offset;
+static u32 s_output_address;
+static u32 s_length;
+static bool s_decrypt;
+
+// Used to notify emulated software after executing command.
+// Pointers don't work with savestates, so CoreTiming events are used instead
+static int s_callback_event_type;
+
+// The following time variables are only used for logging
+static u64 s_realtime_started_us;
+static u64 s_realtime_done_us;
+
+void Start()
+{
+	s_finish_read = CoreTiming::RegisterEvent("FinishReadDVDThread", FinishRead);
+	_assert_(!s_DVD_thread.joinable());
+	s_DVD_thread = std::thread(DVDThread);
+}
+
+void Stop()
+{
+	_assert_(s_DVD_thread.joinable());
+
+	// The DVD thread will return if s_DVD_thread_exiting
+	// is set when it starts working
+	s_DVD_thread_exiting.Set();
+	s_DVD_thread_start_working.Set();
+
+	s_DVD_thread.join();
+
+	s_DVD_thread_exiting.Clear();
+}
+
+void DoState(PointerWrap &p)
+{
+	WaitUntilIdle();
+
+	// TODO: Savestates can be smaller if s_DVD_buffer is not saved
+	p.Do(s_DVD_buffer);
+	p.Do(s_time_read_started);
+	p.Do(s_DVD_success);
+
+	p.Do(s_DVD_offset);
+	p.Do(s_output_address);
+	p.Do(s_length);
+	p.Do(s_decrypt);
+	p.Do(s_callback_event_type);
+
+	// s_realtime_started_us and s_realtime_done_us aren't savestated
+	// because they rely on the current system's time.
+	// This means that loading a savestate might cause
+	// incorrect times to be logged once.
+}
+
+void WaitUntilIdle()
+{
+	_assert_(Core::IsCPUThread());
```

Comment:
This assert is hit with savestates. They aren't called from the CPU thread

#8  0x00000000005f1ead in DVDThread::WaitUntilIdle () at ../Source/Core/Core/HW/DVDThread.cpp:105
#9  0x00000000005f2389 in DVDThread::DoState (p=...) at ../Source/Core/Core/HW/DVDThread.cpp:83
#10 0x0000000000601783 in HW::DoState (p=...) at ../Source/Core/Core/HW/HW.cpp:97
#11 0x00000000005c5444 in State::DoState (p=...) at ../Source/Core/Core/State.cpp:172
#12 0x00000000005c6bf4 in State::SaveToBuffer (buffer=std::vector of length 0, capacity 0) at ../Source/Core/Core/State.cpp:204
#13 0x00000000005c7bc7 in State::LoadAs (filename="/home/markus/.dolphin-emu/StateSaves/GE4E7D.s01") at ../Source/Core/Core/State.cpp:518
#14 0x00000000005c807a in State::LoadAs (filename="/home/markus/.dolphin-emu/StateSaves/GE4E7D.s01") at ../Source/Core/Core/State.cpp:641
#15 State::Load (slot=slot@entry=1) at ../Source/Core/Core/State.cpp:640
#16 0x000000000050bf1b in CFrame::ParseHotkeys (this=0x18f3cc0) at ../Source/Core/DolphinWX/Frame.cpp:1513
#17 0x00007ffff21d393e in wxAppConsoleBase::CallEventHandler(wxEvtHandler*, wxEventFunctor&, wxEvent&) const () from /usr/lib/libwx_baseu-3.0.so.0
#18 0x00007ffff236b2e8 in wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#19 0x00007ffff236b6de in wxEvtHandler::SearchDynamicEventTable(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#20 0x00007ffff236b76e in wxEvtHandler::TryHereOnly(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#21 0x00007ffff236b833 in wxEvtHandler::ProcessEventLocally(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#22 0x00007ffff236b895 in wxEvtHandler::ProcessEvent(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#23 0x00007ffff236b5f7 in wxEvtHandler::SafelyProcessEvent(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#24 0x00007ffff22cb071 in wxTimerImpl::SendEvent() () from /usr/lib/libwx_baseu-3.0.so.0
#25 0x00007ffff2f3e1ab in ?? () from /usr/lib/libwx_gtk2u_core-3.0.so.0
#26 0x00007ffff58526e3 in ?? () from /usr/lib/libglib-2.0.so.0
#27 0x00007ffff5851c7a in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
#28 0x00007ffff5852020 in ?? () from /usr/lib/libglib-2.0.so.0
#29 0x00007ffff5852342 in g_main_loop_run () from /usr/lib/libglib-2.0.so.0
#30 0x00007ffff3676467 in gtk_main () from /usr/lib/libgtk-x11-2.0.so.0
#31 0x00007ffff2f1f485 in wxGUIEventLoop::DoRun() () from /usr/lib/libwx_gtk2u_core-3.0.so.0
#32 0x00007ffff2218c53 in wxEventLoopBase::Run() () from /usr/lib/libwx_baseu-3.0.so.0
#33 0x00007ffff21d7b46 in wxAppConsoleBase::MainLoop() () from /usr/lib/libwx_baseu-3.0.so.0
#34 0x00007ffff226e8f0 in wxEntry(int&, wchar_t**) () from /usr/lib/libwx_baseu-3.0.so.0
#35 0x000000000043e268 in main (argc=1, argv=<optimized out>) at ../Source/Core/DolphinWX/Main.cpp:80